### DISCUSSION TITLE
AlpineJS + Elementary

### DISCUSSION THREAD

**Discussion by amirsaam** [elementary-swift/elementary#39] [Ideas]
I really love the Swift-ish syntax of Elementary and Elementary-HTMX, in my research for the best stack to be used along side Vapor, I found that since HTMX is mostly for talking with server in HTML, but many things do not need this interaction but still many divs need to become reactive and JS-devs made a good solution for it as you may encounter it too, the [AlpineJS](https://github.com/alpinejs/alpine), it may be big request but it may help the ecosystem to make Vapor+Elementary (with HTMX & AlpineJS) to become more reactive and dynamic web stack.

**Comment by amirsaam** [comment]
This way I do think we can utilise UI libraries like [Pines](https://devdojo.com/pines) that are based on Alpine and Tailwind and do not need server-side JS rendering like React libraries.

**Reply by sliemeobn** [reply]
my intention with elementary was to create the bare-bones package that can then be extended or built upon in other packages (ie: build your own component library on top)

it would be fantastic to see people experimenting how to create nice, bundled up libraries with batteries-included components!

**Reply by amirsaam** [reply]
Surely, we don't need to have recreate Pines in Swift (I think?), only if we have AlpineJS package just like HTMX of what you did, Pines is an import and use library with div classes but it needs Alpine tags to run.

**Comment by sliemeobn** [comment]
I totally agree and I am on the same mission!
alpine is a great little tool to throw in the mix for light-weight client-side interactivity.

what I am looking into for my own project is using web component libraries, and [shoelace](https://shoelace.style/) looks really nice.
I think rendering out the "big page" with server-side swift, using HTMX for laser-focused server interactions, and shoelace (or web components in general) can get you all the way for a lot of web apps.

also, I have it on my list to tinker with an elementary-based reactive DOM package that runs as WASM in the browser, but that is no small task...

**Reply by amirsaam** [reply]
TBH it's my first time hearing about Shoelace, and doesn't it a React library (based on what I've read on their website)? Are you take apart your front and back containers and creating each on their own isolated container?
If yes my proposal was to have Vapor as backend the frontend on same container with this stack cuz looks like they can be on same container cuz of no server-side JS rendering unlike React frameworks.

**Reply by sliemeobn** [reply]
shoelace has nothing to do with React, it's pre-packaged components that in any HTML page (using web components). 
a bit like if you took a snippet of Pines and packaged it up in a nice custom HTML tag.

**Reply by amirsaam** [reply]
Oh, the "Framework" section draw my attention that's why I thought it's a React thing, interesting.
Me myself am a more Radix appreciator that's why I like Pine more.

**Comment by amirsaam** [comment]
VATHe Stack: Vapor+Alpine+Tailwind+HTMX (fenced area, in which the small cattle is herded together for the night (lol))
or
HATHe Stack: Hummingbird+Alpine+Tailwind+HTMX (third-person singular simple present indicative of [have])

(the lowercase "e" stands for "Elementary")

**Reply by sliemeobn** [reply]
I'm more of a hummingbird guy myself, but sure ; )

**Reply by amirsaam** [reply]
haha, it could be any first word ;)

**Comment by amirsaam** [comment]
No AlpineElementary?

**Reply by sliemeobn** [reply]
not yet, I use it in a project with elementary, but polishing up the API to make sense as a releasable package requires a bit of thinking.

also, I have been thinking about setting up a github org for elementary to collect all the packages under one umbrella. this could make the addition of community packages easier. but I am a bit short on playtime at the moment : /

**Reply by amirsaam** [reply]
I'm looking forward towards it no matter how long it does take <3

For the github org, it makes sense absolutely!

**Reply by erikbdev** [reply]
> not yet, I use it in a project with elementary, but polishing up the API to make sense as a releasable package requires a bit of thinking.

@sliemeobn  Are there any updates for Elementary + AlpineJS integration? 
 
PS: Elementary is awesome :D

**Reply by sliemeobn** [reply]
sorry, at least none from my side.

I am a bit short on play time at the moment, but I am happy to collaborate if somebody wants to give it a go.

**Reply by amirsaam** [reply]
@sliemeobn hey buddy, can you please share the alpine you use in your project "as is" maybe in your personal account so let us think and help on the polishing part? maybe we could help on it.


PS: I do know the ElementaryUI is taking a lot of your time and effort also congrats on it, it's a promising project that fills the gap between the SwiftUI developers with the web but still, there are many use cases that still need old-fashioned ways to get done.

**Reply by sliemeobn** [reply]
I am sorry to report: I have nothing to share, and I will not work on this in the next weeks / months : /
stilly happy to collaborate on a "community package" or similar.

**Reply by amirsaam** [reply]
> not yet, I use it in a project with elementary, but polishing up the API to make sense as a releasable package requires a bit of thinking.

@sliemeobn i was referring to this part of our conversation, it makes us understand how you used it and how we can utilise it then go to the path to make a valid api. but if it no longer available, then it's something else. Tbh my codes are not as polished as your works, but I am ready to do a community package with you under the banner.
If you be able to make a roadmap and a protocol on how it should be done, then I'll try to make it as far as I can.

**Reply by sliemeobn** [reply]
sorry amir, I hate to disappoint, but I have nothing to give you at the moment : (

_>> If you be able to make a roadmap and a protocol on how it should be done, then I'll try to make it as far as I can._
well, that's the hard part, isn't it.

I'd start with something like this (see below) - but how to really shape the API to make it "worthy" of being a package, and to actually be helpful and carry it's own weight, well that's the part that will require research, playing with various shapes, taking decisions, and so on. I don't have the bandwidth for this ATM.

```swift
public extension HTMLAttribute where Tag: HTMLTrait.Attributes.Global {
    enum alpine {}
}

public extension HTMLAttribute.alpine {
    static func data(_ code: String) -> HTMLAttribute {
        HTMLAttribute(name: "x-data", value: code)
    }
}
```

**Reply by amirsaam** [reply]
> sorry amir, I hate to disappoint, but I have nothing to give you at the moment : (

there is nothing for you to be sorry about mate, you've done a great favour so far

> I'd start with something like this (see below)

thanks for sharing it, i'll try to be part of this once we got a stable internet connection in iran since if you know what's going on around here 😅

**Reply by sliemeobn** [reply]
oh my... all the best to you. I hope the suffering and misery will end as soon as possible : (

**Reply by amirsaam** [reply]
@sliemeobn [ElementaryAlpine](https://github.com/amirsaam/elementary-alpine) 💃 
If you accept, I would like to give the ownership to Elementary.

**Reply by sliemeobn** [reply]
look cool, thanks!
I'll take a closer look later this week. maybe we could set up a "elementary-swift-community" org or similar.
but at the moment I don't have the bandwidth to onboard this, sorry

**Reply by amirsaam** [reply]
no pressure buddy, it was just an idea 😅
also i plan to make a Pines UI after Alpine, it would be nice to have a community org, indeed.

**Comment by balthild** [comment]
Alpine is rather simple and is easy to implement in elementary. Here's a snippet:

```swift

extension HTMLAttribute {
	/// A namespace for Alpine.js attributes.
	/// See the [Alpine.js docs](https://alpinejs.dev) for more information.
	public enum x {}
}

extension HTMLAttribute.x {
	@inlinable
	public static func data(_ expr: String) -> HTMLAttribute {
		.init(name: "x-data", value: expr)
	}

	@inlinable
	public static func bind(_ name: String, _ expr: String) -> HTMLAttribute {
		.init(name: "x-bind:\(name)", value: expr)
	}

	@inlinable
	public static func on(_ event: String, _ expr: String) -> HTMLAttribute {
		.init(name: "x-on:\(event)", value: expr)
	}

	@inlinable
	public static func text(_ expr: String) -> HTMLAttribute {
		.init(name: "x-text", value: expr)
	}

	@inlinable
	public static func html(_ expr: String) -> HTMLAttribute {
		.init(name: "x-html", value: expr)
	}

	@inlinable
	public static func model(_ expr: String) -> HTMLAttribute {
		.init(name: "x-model", value: expr)
	}

	@inlinable
	public static func modelable(_ expr: String) -> HTMLAttribute {
		.init(name: "x-modelable", value: expr)
	}

	@inlinable
	public static func show(_ expr: String) -> HTMLAttribute {
		.init(name: "x-show", value: expr)
	}

	@inlinable
	public static func transition(_ expr: String) -> HTMLAttribute {
		.init(name: "x-transition", value: expr)
	}

	@inlinable
	public static func `for`(_ expr: String) -> HTMLAttribute {
		.init(name: "x-for", value: expr)
	}

	@inlinable
	public static func `if`(_ expr: String) -> HTMLAttribute {
		.init(name: "x-if", value: expr)
	}

	@inlinable
	public static func `init`(_ expr: String) -> HTMLAttribute {
		.init(name: "x-init", value: expr)
	}

	@inlinable
	public static func effect(_ expr: String) -> HTMLAttribute {
		.init(name: "x-effect", value: expr)
	}

	@inlinable
	public static func ref(_ expr: String) -> HTMLAttribute {
		.init(name: "x-ref", value: expr)
	}

	@inlinable
	public static func id(_ expr: String) -> HTMLAttribute {
		.init(name: "x-id", value: expr)
	}

	@inlinable
	public static func teleport(_ expr: String) -> HTMLAttribute {
		.init(name: "x-teleport", value: expr)
	}

	@inlinable
	public static var cloak: HTMLAttribute {
		.init(name: "x-cloak", value: nil)
	}

	@inlinable
	public static var ignore: HTMLAttribute {
		.init(name: "x-ignore", value: nil)
	}
}
```

**Reply by amirsaam** [reply]
you can already check out my implementation that I've mentioned before here https://github.com/amirsaam/elementary-alpine
